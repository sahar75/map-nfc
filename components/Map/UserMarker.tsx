import { useUserLocationStore } from "../../store/userLocation";
import * as Location from "expo-location";
import React, { useCallback, useEffect, useState } from "react";
import { Alert } from "react-native";
import { Marker } from "react-native-maps";
const locationImg = require("../../assets/images/user-location.png");

interface IProps {
  selectedRoutePoints: {
    lat: number;
    lng: number;
  }[];
}

const UserMarker: React.FC<IProps> = ({ selectedRoutePoints }) => {
  const { userLocation, setUserLocation } = useUserLocationStore();
  const [userMarkerRotation, setUserMarkerRotation] = useState(0);

  const handleLocationUpdate = useCallback(
    (location: Location.LocationObject) => {
      if (location.coords) {
        const newUserLocation = {
          latitude: location.coords.latitude,
          longitude: location.coords.longitude,
        };

        if (userLocation) {
          const angle =
            Math.atan2(
              newUserLocation.longitude - userLocation.longitude,
              newUserLocation.latitude - userLocation.latitude
            ) *
            (180 / Math.PI);

          setUserMarkerRotation(-angle + 90);
        }
        setUserLocation(newUserLocation);
      }
    },
    [selectedRoutePoints, userLocation]
  );

  useEffect(() => {
    const requestLocationPermissions = async () => {
      let { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== "granted") {
        Alert.alert(
          "Permission Denied",
          "Permission to access location was denied"
        );
        return;
      }

      const locationSubscription = Location.watchPositionAsync(
        {
          accuracy: Location.Accuracy.BestForNavigation,
          distanceInterval: 1,
          timeInterval: 3000,
        },
        handleLocationUpdate
      );

      return () => locationSubscription.then((sub) => sub.remove());
    };
    requestLocationPermissions();
  }, []);

  return (
    <Marker
      coordinate={userLocation}
      rotation={userMarkerRotation}
      zIndex={100}
      image={locationImg}
    />
  );
};

export default React.memo(UserMarker);
