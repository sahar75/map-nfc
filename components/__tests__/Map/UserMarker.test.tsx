import React from "react";
import TestRenderer from "react-test-renderer";
import UserMarker from "../../Map/UserMarker";

// Mock the Marker component from react-native-maps
jest.mock("react-native-maps", () => {
  return {
    Marker: "Marker",
  };
});

// Mock expo-location
jest.mock("expo-location", () => ({
  requestForegroundPermissionsAsync: jest
    .fn()
    .mockResolvedValue({ status: "granted" }),
  watchPositionAsync: jest.fn().mockResolvedValue({ remove: jest.fn() }),
  Accuracy: {
    BestForNavigation: "BestForNavigation",
  },
}));

// Mock the useUserLocationStore hook
jest.mock("../../../store/userLocation", () => ({
  useUserLocationStore: jest.fn().mockReturnValue({
    userLocation: { latitude: 37.78825, longitude: -122.4324 },
    setUserLocation: jest.fn(),
  }),
}));

describe("UserMarker", () => {
  it("should render correctly with user location", () => {
    const tree = TestRenderer.create(
      <UserMarker selectedRoutePoints={[]} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should handle location updates correctly", () => {
    // Create an instance to check if location update logic runs
    TestRenderer.create(<UserMarker selectedRoutePoints={[]} />);
    // Here you could test specific behaviors or state changes if needed
    // This might include checking if the mocked functions were called
  });
});
