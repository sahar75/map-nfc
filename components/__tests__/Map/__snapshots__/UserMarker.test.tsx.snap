// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMarker should render correctly with user location 1`] = `
<Marker
  coordinate={
    {
      "latitude": 37.78825,
      "longitude": -122.4324,
    }
  }
  image={
    {
      "testUri": "../../../assets/images/user-location.png",
    }
  }
  rotation={0}
  zIndex={100}
/>
`;
